api-docs: Clarify error response keys and handling documentation.

laurynmm opened the pull request:
**Notes on changes**:
- Updates the main description of the API error handling article so that the `code` key in the error reponse is more accurate.
 - Adds a bullet point to the API changelog for feature level 76 about the `code` key being consistently in all error responses.
   - This is based on commit 8b9011dff8, which completed the migration to JsonableError, at which point the API feature level was already set to 75, and not yet set to 76.
   - Alternatively, we could just use the major release (Zulip 5.0) for when clients can expect all error responses to contain a `code` key.
 - Also, revises the other error related feature level 76 entries in the API changelog.
 - Updates the descriptions of the shared error responses documented on the error handling page to include information about any extra fields and to have the changes notes at the end of the description.

**Relevant CZO conversations**:
- [#api design > error docs](https://chat.zulip.org/#narrow/stream/378-api-design/topic/error.20docs/near/1308992)
- [#api documentation > error responses](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/documenting.20error.20responses/near/1645188)

**Screenshots and screen captures:**

<details>
<summary>API changelog feature level 76 updated entry</summary>

[Current documentation](https://zulip.com/api/changelog) - find on page "feature level 76"
![Screenshot from 2024-06-17 17-24-43](https://github.com/zulip/zulip/assets/63245456/3e837487-f185-460f-998c-1922c3b248e3)
</details>
<details>
<summary>Error handling article</summary>

[Current documentation](https://zulip.com/api/rest-error-handling)
![Screenshot from 2024-06-17 17-25-02](https://github.com/zulip/zulip/assets/63245456/6148f919-4bb0-4f56-83ff-518a0bc18055)
![Screenshot from 2024-06-17 17-25-15](https://github.com/zulip/zulip/assets/63245456/61d1147d-c977-482a-89cb-b8177bb7f502)
![Screenshot from 2024-06-17 17-25-31](https://github.com/zulip/zulip/assets/63245456/43f72bdb-677e-4424-831e-5992c508ce2b)
</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on api_docs/rest-error-handling.md:
```diff
@@ -2,20 +2,40 @@
+
+As of Zulip 5.0 (feature level 76), all error responses also contain a
+`code` key with a machine-readable error string. The default value for
+`code` is `"BAD_REQUEST"`, which is returned for a number of errors.
```
Probably this should be explained and formatted as a mandatory key like the previous two, and the pre-76 discussion should be marked as a "Changes" entry.

laurynmm replied:
Done in updated version.

timabbott commented on api_docs/changelog.md:
```diff
@@ -1689,11 +1689,18 @@ No changes; feature level used for Zulip 5.0 release.
+  now include the `code` key with a machine-readable string value. The
+  default value for this key is `"BAD_REQUEST"`. As of this feature level,
+  clients should check the `code` key when looking for specific error
+  conditions.
```
I think we always had the opinion that you should look at `code` to identify specific error conditions; what changed is that we started providing a standard `BAD_REQUEST` value for all API responses, rather than encoding "we haven't allocated a specific code to this failure" via absence of a `code`.

laurynmm replied:
Make sense. Updated as such.

laurynmm commented:
**Updates**:
- Updated for review feedback above:
  - The `code` key is documented like the other mandatory keys (`msg` and `result`)
  - The new feature level 76 API changelog entry to just note the `code` key always being present and default value.
- Added a header for the common error response section so that the **Changes** entry for the main description reads more clearly. It's also nice to have a header to link to for this section in the API changelog.
- Updated API changelog feature level 76 to have "Error handling" to start the entry vs "All REST API endpoints" as that seemed clearer.
  - I left the API changelog entry for feature level 167 (about the `ignored_parameters_unsupported`) to start with "All REST API endpoints" even though it links to the error handling page because that seemed a better fit there.
- Added a commit for feature level 36 rate limit error documentation, which really just updates those to better match our current conventions. See screenshots below.

**Updated screenshots**:

<details>
<summary>API changelog updates</summary>

[Current documentation](https://zulip.com/api/changelog)
![Screenshot from 2024-06-18 14-07-41](https://github.com/zulip/zulip/assets/63245456/4c9c0226-e80f-43ef-85ec-61b828f62d93)
![Screenshot from 2024-06-18 14-08-00](https://github.com/zulip/zulip/assets/63245456/35ee7794-4f48-4851-96f5-970dacb3116f)
</details>
<details>
<summary>Error handling documentation</summary>

[Current documentation](https://zulip.com/api/rest-error-handling)
![Screenshot from 2024-06-18 14-08-47](https://github.com/zulip/zulip/assets/63245456/6f152189-3088-407a-bee0-9a693e4c32b7)
![Screenshot from 2024-06-18 14-09-02](https://github.com/zulip/zulip/assets/63245456/3861233e-a50f-43b5-9ab1-03c3326d7893)
![Screenshot from 2024-06-18 14-09-14](https://github.com/zulip/zulip/assets/63245456/34b06ca7-a4a9-4d9e-8e51-e67f9d3e89d0)
</details>

timabbott commented:
Merged, thanks @laurynmm!
